src/uct/ib/base/ib_device.c

```diff
@@ -629,11 +629,37 @@ ucs_status_t uct_ib_device_query(uct_ib_device_t *dev,
     return UCS_OK;
 }
 
+static int uct_ib_device_is_iwarp(uct_ib_device_t *dev)
+{
+    return dev->ibv_context->device->transport_type == IBV_TRANSPORT_IWARP;
+}
+
 ucs_status_t uct_ib_device_init(uct_ib_device_t *dev,
                                 struct ibv_device *ibv_device, int async_events
                                 UCS_STATS_ARG(ucs_stats_node_t *stats_parent))
 {
     ucs_status_t status;
+    uint8_t port_num;
+    uint8_t usable_port_cnt = 0;
+
+    if (uct_ib_device_is_iwarp(dev)) {
```

Comment:
IMO we should modify https://github.com/openucx/ucx/blob/a767face90f883dc3de61a9b275bc82b42dacda4/src/uct/ib/base/ib_device.c#L777-L787

1. move uct_ib_device_is_iwarp() from uct_ib_device_port_check to https://github.com/openucx/ucx/blob/ee5fa05ac5fdcacda190ec6f95a2c00475e6008d/src/uct/ib/base/ib_md.c#L1345
2. check that port is RoCE or IB in https://github.com/openucx/ucx/blob/a767face90f883dc3de61a9b275bc82b42dacda4/src/uct/ib/base/ib_device.c#L783

Reply:
I agree that the solution does not look super clean. The logic between MD creating and the rest of the code is not unified. I was really puzzled when I went through the code.

Reply:
@yosefe I implemented what you suggested and it does not look good.
1 - Will work but ...
2 - It is not going to work because `uct_ib_device_port_check` that you referenced is invoked after MD creation, which too late.

Just addressing #1 is possible but then you only execute the iwarp part of the check in `uct_ib_device_is_accessible` and the second half (not roce or IB) in uct_ib_device_init which makes code much more complex to understand. Essentially the logical check "not verbs compliant device that has verbs" gets spread over different places. IMHO current implementation is more reasonable.

Reply:
we are checking the port type after creating the md because in case some ports are supported and some are not, it makes more sense to disable only the unsupported ports and not the entire device.
if the device does not have any supported ports we would eventually close the md anyway.

Reply:
> @yosefe I understand why you are checking port by port and the patch is doing the same in much earlier stage so we don't create MD for "non-verbs" devices.

@shamisp what is the reason we must avoid creating such MD?

Reply:
@yosefe Because the TL does not work on transport layers other than `IBV_TRANSPORT_IB`, which cover both ROCE and IB.
We know that  IBV_TRANSPORT_USNIC, IBV_TRANSPORT_USNIC_UDP, and IBV_TRANSPORT_UNSPECIFIED. So, why to enable and then cause failures and leaks when on such systems only TCP has to be used in default mode.

Reply:
LGTM